sdk/iot/hub/src/az_iot_hub_client_twin.c

```diff
@@ -103,16 +119,118 @@ AZ_NODISCARD az_result az_iot_hub_client_twin_patch_publish_topic_get(
+  int32_t required_length = az_span_size(az_iot_hub_twin_topic_prefix)
+      + az_span_size(az_iot_hub_twin_patch_pub_topic)
+      + (int32_t)sizeof(az_iot_hub_client_twin_question)
+      + az_span_size(az_iot_hub_client_request_id_span)
+      + (int32_t)sizeof(az_iot_hub_client_twin_equals) + az_span_size(request_id);
 
   AZ_RETURN_IF_NOT_ENOUGH_SIZE(mqtt_topic, required_length);
 
-  az_span remainder = az_span_copy(mqtt_topic, az_iot_hub_twin_patch_pub_topic);
-  remainder = az_span_copy(remainder, az_iot_hub_client_twin_request_id_suffix);
-  az_span_copy(remainder, request_id);
+  az_span remainder = az_span_copy(mqtt_topic, az_iot_hub_twin_topic_prefix);
+  remainder = az_span_copy(remainder, az_iot_hub_twin_patch_pub_topic);
+  remainder = az_span_copy_u8(remainder, az_iot_hub_client_twin_question);
+  remainder = az_span_copy(remainder, az_iot_hub_client_request_id_span);
+  remainder = az_span_copy_u8(remainder, az_iot_hub_client_twin_equals);
+  remainder = az_span_copy(remainder, request_id);
 
   *out_mqtt_topic = az_span_slice(mqtt_topic, 0, required_length);
 
   return AZ_OK;
 }
+
+AZ_NODISCARD az_result az_iot_hub_client_twin_received_topic_parse(
+    az_iot_hub_client const* client,
+    az_span received_topic,
+    az_iot_hub_client_twin_response* out_twin_response)
+{
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(received_topic, 1, false);
+  AZ_PRECONDITION_NOT_NULL(out_twin_response);
+
+  az_result result;
+
+  int32_t twin_index;
+  // Check if is related to twin or not
+  if ((twin_index = az_span_find(received_topic, az_iot_hub_twin_topic_prefix)) >= 0)
+  {
+    int32_t twin_feature_index;
+    az_span twin_feature_span
+        = az_span_slice(received_topic, twin_index, az_span_size(received_topic));
+
+    if ((twin_feature_index = az_span_find(twin_feature_span, az_iot_hub_twin_response_sub_topic))
+        >= 0)
+    {
+      // Is a res case
+      az_span remainder;
+      az_span status_str = az_span_token(
+          az_span_slice(
+              received_topic,
+              twin_feature_index + az_span_size(az_iot_hub_twin_response_sub_topic),
+              az_span_size(received_topic)),
+          AZ_SPAN_FROM_STR("/"),
+          &remainder);
+
+      // Get status and convert to enum
+      uint32_t status_int;
+      AZ_RETURN_IF_FAILED(az_span_atou32(status_str, &status_int));
+      AZ_RETURN_IF_FAILED(az_iot_get_status_from_uint32(status_int, &out_twin_response->status));
+
+      // Get request id prop value
+      az_iot_hub_client_properties props;
+      az_span prop_span = az_span_slice(remainder, 1, az_span_size(remainder));
+      AZ_RETURN_IF_FAILED(
+          az_iot_hub_client_properties_init(&props, prop_span, az_span_size(prop_span)));
+      AZ_RETURN_IF_FAILED(az_iot_hub_client_properties_find(
+          &props, az_iot_hub_client_request_id_span, &out_twin_response->request_id));
+
+      if (out_twin_response->status == AZ_IOT_STATUS_NO_CONTENT)
```

Comment:
Caveat here for those reviewing: We currently can't map an error return value on the `res/#` to it's source. Meaning if we get a `$iothub/twin/res/400/` here, we can't natively tell if it was from a reported prop request or a get twin request because we don't have a queue of messages to resolve the request ID with. Proposal here could be to have a fourth `out_twin_response->response_type` here to designate it an error. Something like `AZ_IOT_CLIENT_TWIN_RESPONSE_TYPE_ERROR`.

Reply:
On the other hand, we could leave that part of the parsing to the implementing layer, documenting it well. IT would be the most efficient way, although slightly more complicated for the customer to implement.